### PR TITLE
Add custom repos before installing rpm packages

### PR DESCRIPTION
**Pull request by kshtsk**
It is a set of patches from suse/teuthology related to packaging and --test-repo feature.

The idea of this feature is to bypass buildpackages phase and let teuthology install ceph packages from the specified test repos in 'install' task. For openstack it speeds up bootstrapping a lot, first of all it does not spend tons time for building ceph packages, and does not need to allocate automatically builder and packages servers in cloud services which needs extra cost.

Signed-off-by: Nathan Cutler <ncutler@suse.com>
Signed-off-by: Kyr Shatskyy <kshatskyy@suse.com>

### CODE REVIEW

**Comment by kshtsk** [comment]
retest this please

File: teuthology/orchestra/opsys.py [hunk 1]
```diff
@@ -85,7 +86,7 @@ def __init__(self, name=None, version=None, codename=None):
     @staticmethod
     def _version_to_codename(name, version):
         for (_version, codename) in DISTRO_CODENAME_MAP[name].iteritems():
-            if version == _version or version.split('.')[0] == _version:
+            if str(version) == _version or str(version).split('.')[0] == _version:
```
Comment (gregsfortytwo) [inline teuthology/orchestra/opsys.py]: Why the change to stringify variables here? Aren’t they already strings?

File: teuthology/packaging.py [hunk 2]
```diff
@@ -641,7 +641,7 @@ def _get_version(self):
         if not version:
             version = DEFAULT_OS_VERSION.get(self.os_type)
 
-        return version
+        return str(version)
```
Comment (gregsfortytwo) [inline teuthology/packaging.py]: And here.

File: teuthology/run.py [hunk 3]
```diff
@@ -168,10 +168,15 @@ def validate_tasks(config):
 
 
 def get_initial_tasks(lock, config, machine_type):
-    init_tasks = [
-        {'internal.check_packages': None},
-        {'internal.buildpackages_prep': None},
-    ]
+    init_tasks = []
+    overrides = config.get('overrides')
+    having_repos = ('install' in config and 'repos' in config.get('install')) \
```
Comment (gregsfortytwo) [inline teuthology/run.py]: I don’t think I understand what’s happening here. I presume you want to bypass these internal tasks if not in the upstream lab, but the conditions don’t make a lot of sense to me.

File: teuthology/orchestra/opsys.py (hunk 1 above)
Comment (kshtsk) [reply teuthology/orchestra/opsys.py]: > Why the change to stringify variables here? Aren’t they already strings?

At the time of patch applied, the version looks like 'X.Y' and it appeared here as floating, so... this code made sure that there is a string.

File: teuthology/packaging.py (hunk 2 above)
Comment (kshtsk) [reply teuthology/packaging.py]: Same as above.

File: teuthology/run.py (hunk 3 above)
Comment (kshtsk) [reply teuthology/run.py]: Yes, the idea we don't want to build ceph packages ourself if we provided test repos.
